query-parser/src/parse.rs

```diff
@@ -402,17 +431,28 @@ def_parser!(Find, query, FindQuery, {
                 match part {
                     FindQueryPart::FindSpec(x) => find_spec = Some(x),
                     FindQueryPart::In(x) => in_vars = Some(x),
+                    FindQueryPart::Limit(x) => limit = x,
                     FindQueryPart::Order(x) => order_clauses = Some(x),
                     FindQueryPart::WhereClauses(x) => where_clauses = Some(x),
                     FindQueryPart::With(x) => with_vars = Some(x),
                 }
             }
 
+            // Make sure that if we have `:limit ?x`, `?x` appears in `:in`.
```

Comment:
I opted for a kind of internal consistency: everything between those `[]` is in the same namespace of variables, is bound in the same way, and so on.

Only a step away from that is ensuring that a limit isn't used elsewhere within the query, as you suggest — yes, it might well be human error.

But that's an additional bit of metadata that would need to be passed around, an additional bit of validation, and it raises good questions: can I project the limit from the query to allow for pipelined processing of results, even if I'm not allowed to use a limit var inside `:where`?

There are also legitimate reasons why a user might want to use a limit as part of `:where`:

```
[:find ?x ?person
 :in ?limit
 :where
 [?x :foo/knows ?person]
 (or-join [?person]
   [?person :foo/employment "faculty"]
   (> ?limit 10))
 :limit ?limit]
```

— that is, find _everyone_ `?x` knows if we're fetching more than 10, but only include faculty if we're fetching 10 or fewer. You can make a similar case for including limits inside `:order` clauses to control ordering based on how many results will be returned.